File: ai-code-smell-linter/aicslint/parser.py
"""Multi-language AST parser using tree-sitter."""

import tree_sitter_python as tspython
import tree_sitter_javascript as tsjavascript
import tree_sitter_typescript as tstypescript
from tree_sitter import Language, Parser

PY_LANGUAGE = Language(tspython.language())
JS_LANGUAGE = Language(tsjavascript.language())
TS_LANGUAGE = Language(tstypescript.language_typescript())

LANG_MAP = {
    ".py": PY_LANGUAGE,
    ".js": JS_LANGUAGE,
    ".ts": TS_LANGUAGE,
    ".tsx": TS_LANGUAGE,
    ".jsx": JS_LANGUAGE,
}

LANG_NAME_MAP = {
    ".py": "python",
    ".js": "javascript",
    ".ts": "typescript",
    ".tsx": "typescript",
    ".jsx": "javascript",
}


def get_language(ext: str) -> Language | None:
    return LANG_MAP.get(ext)


def get_language_name(ext: str) -> str | None:
    return LANG_NAME_MAP.get(ext)


def parse_code(source: str, ext: str) -> tuple:
    """Parse source code and return (tree, language_name) or (None, None)."""
    lang = get_language(ext)
    if lang is None:
        return None, None
    parser = Parser(lang)
    tree = parser.parse(source.encode("utf-8"))
    return tree, get_language_name(ext)

File: dark-path-detector/src/parser.js
const Parser = require('tree-sitter');
const JavaScript = require('tree-sitter-javascript');
const TypeScript = require('tree-sitter-typescript').typescript;
const TSX = require('tree-sitter-typescript').tsx;

/**
 * Create a parser for the given file extension.
 * Returns null if the extension is not supported.
 */
function createParser(ext) {
  const parser = new Parser();
  switch (ext) {
    case '.js':
    case '.mjs':
    case '.cjs':
      parser.setLanguage(JavaScript);
      return parser;
    case '.ts':
      parser.setLanguage(TypeScript);
      return parser;
    case '.tsx':
    case '.jsx':
      parser.setLanguage(TSX);
      return parser;
    default:
      return null;
  }
}

/**
 * Parse source code and return the tree.
 */
function parseSource(source, ext) {
  const parser = createParser(ext);
  if (!parser) return null;
  return parser.parse(source);
}

module.exports = { createParser, parseSource };
